module/mempool/consensus/execution_tree.go

```diff
@@ -0,0 +1,216 @@
+// of the block the receipt is for. We enforce data consistency on an API
+// level by using the block header as input.
+func (et *ExecutionTree) AddReceipt(receipt *flow.ExecutionReceipt, block *flow.Header) (bool, error) {
+	et.Lock()
+	defer et.Unlock()
+
+	// drop receipts for block heights lower than the lowest height.
+	if block.Height < et.forest.LowestLevel {
+		return false, nil
+	}
+
+	// sanity check: initial result should be for block
+	if block.ID() != receipt.ExecutionResult.BlockID {
+		return false, fmt.Errorf("receipt is for different block")
+	}
+
+	receiptsForResult, err := et.getEquivalenceClass(&receipt.ExecutionResult, block)
+	if err != nil {
+		return false, fmt.Errorf("failed to get equivalence class for result (%x): %w", receipt.ExecutionResult.ID(), err)
+	}
+
+	added, err := receiptsForResult.AddReceipt(receipt)
+	if err != nil {
+		return false, fmt.Errorf("failed to add receipt to its equivalence class: %w", err)
+	}
+	et.size += added
+	return added > 0, nil
+}
+
+// ReachableReceipts returns a slice of ExecutionReceipt, whose result
+// is computationally reachable from resultID. Context:
+//  * Conceptually, the Execution results form a tree, which we refer to as
+//    Execution Tree. A fork in the execution can be due to a fork in the main
+//    chain. Furthermore, the execution forks if ENs disagree about the result
+//    for the same block.
+//  * As the ID of an execution result contains the BlockID, which the result
+//    for, all Execution Results with the same ID necessarily are for the same
+//    block. All Execution Receipts committing to the same result from an
+//    equivalence class and can be represented as one vertex in the Execution
+//    Tree.
+//  * An execution result r1 points (field ExecutionResult.ParentResultID) to
+//    its parent result r0 , whose end state was used as the starting state
+//    to compute r1. Formally, we have an edge r0 -> r1 in the Execution Tree,
+//    if a result r1 is stored in the mempool, whose ParentResultID points to
+//    r0.
+// ReachableReceipts traverses the Execution Tree from the provided resultID.
+// Execution Receipts are traversed in a parent-first manner, meaning that
+// a receipt committing to the parent result is traversed first _before_
+// the receipt committing to the derived result.
+// The algorithm only traverses to results, for which there exists a
+// sequence of interim result in the mempool without any gaps.
+func (et *ExecutionTree) ReachableReceipts(resultID flow.Identifier, blockFilter mempool.BlockFilter, receiptFilter mempool.ReceiptFilter) ([]*flow.ExecutionReceipt, error) {
+	et.RLock()
+	defer et.RUnlock()
+
+	vertex, found := et.forest.GetVertex(resultID)
+	if !found {
+		return nil, fmt.Errorf("unknown result id %x", resultID)
+	}
+
+	receipts := make([]*flow.ExecutionReceipt, 0, 10) // we expect just below 10 execution Receipts per call
+	et.reachableReceipts(vertex, blockFilter, receiptFilter, &receipts)
```

Comment:
better if the function could return the receipts instead of mutating the input argument